### PR TITLE
Refines raises list in osproc 

### PR DESCRIPTION
**Pull request by ringabout**
closes #19360

### CODE REVIEW

**Comment by ringabout** [comment]
```nim
raiseOSError(lastError,
    "Requested command not found: '$1'. OS error:" % command)
```

`%` raises `ValueError` if `formatspec` is malformed. But it cannot raise in this case, should we add `ValueError` to the raises list of `startProcess`? Or should we suppress the check using `try, except` ? Or should we rewrite it in plain string concatenation, which doesn't raise `ValueError`?

**Comment by arnetheduck** [comment]
> But it cannot raise in this case, should we add ValueError to the raises list of startProcess? 

No - it doesn't make sense in that case, because why should a "process" function raise a `ValueError`? - in our code, we do `try: ... except ValueError: raiseAssert "Cannot happen"` in cases like this, so as to catch programming and refactoring mistakes down the line (such as if someone changes the format string).

Another more subtle point: once you add a `raises` list to the function signature, you can only make it *more constrained* from that point onwards, or you break backwards compatibility just as if you had changed the type of a parameter. As such, it _sometimes_ makes sense to make the `raises` list more "broad" than it in reality is today, anticipating that the function in the future may need it - for that to work reasonably well, you do one of two things:
* develop an exception hierarchy
* "translate" all exceptions to the appropriate abstraction level, ie introduce a `ProcessError` whose parent exception is whatever detail exception there is.

The process of adding raises lists to the std lib is ideally done as consistently and deliberately as anything else that's added to an API - ie it should be deliberate, consistent and structured, and ideally not a random outcome of the current implementation.

**Comment by ringabout** [comment]
```nim
proc waitForExit(p: Process, timeout: int = -1): int =
   if e.tv_sec == posix.Time(0):
    raise newException(ValueError, "System time was modified")
```

`waitForExit` raises `ValueError` on Linux for now, probably it's better to raise `OSError` or something. I'm not sure and that's a breaking change. So I didn't touch them.

File: lib/pure/osproc.nim [hunk 1]
```diff
@@ -746,8 +748,11 @@ when defined(windows) and not defined(useNimRtl):
       const errInvalidParameter = 87.int
       const errFileNotFound = 2.int
       if lastError.int in {errInvalidParameter, errFileNotFound}:
-        raiseOSError(lastError,
-            "Requested command not found: '$1'. OS error:" % command)
+        try:
+          raiseOSError(lastError,
+              "Requested command not found: '$1'. OS error:" % command)
+        except ValueError:
+          discard
```
Comment (Araq) [inline lib/pure/osproc.nim]:
```suggestion
        raiseOSError(lastError,
              "Requested command not found: '" & command & "'. OS error:")
```

File: lib/pure/osproc.nim [hunk 2]
```diff
@@ -1130,9 +1132,12 @@ elif not defined(useNimRtl):
...
       if sizeRead == sizeof(error):
-        raiseOSError(osLastError(),
-                     "Could not find command: '$1'. OS error: $2" %
-                      [$data.sysCommand, $strerror(error)])
+        try:
+          raiseOSError(osLastError(),
+                      "Could not find command: '$1'. OS error: $2" %
+                        [$data.sysCommand, $strerror(error)])
+        except ValueError:
+          discard
```
Comment (Araq) [inline lib/pure/osproc.nim:1140]: Same: Avoid `%` here and use string concats or strformat.

Comment (Araq) [inline lib/pure/osproc.nim]:
```suggestion
        raiseOSError(osLastError(),
                      "Could not find command: '" & $data.sysCommand & "'. OS error: " & $strerror(error))
```

**Comment by ringabout** [comment]
@Araq so `ValueError` needs to be added to raises lists?

**Comment by arnetheduck** [comment]
`cast raises: []` is a pretty unsafe construct - it would indeed be better to remove offending exceptions with `try/except HERE: raiseAssert` which is more future-proof / allows detecting changes in assumptions on the underlying code to be detected earlier - from practical experience, this happens more often than not when people start thinking about error handling in code which previously didn't consider the error part of control flow

**Pull request merged**